### PR TITLE
PoS Implement Bitset data structure

### PR DESCRIPTION
**Pull request by tholonious**

### CODE REVIEW

**Comment by tholonious** [comment]
Current dependencies on/for this PR:
* feature/pos-messaging-schema
  * **PR #555** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/555" target="_blank">Graphite</a> 
    * **PR #558** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/558" target="_blank">Graphite</a> 
      * **PR #564** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/564" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/564?utm_source=stack-comment).

File: utils/bitset/bitset.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
+package bitset
+
+import (
+	"math/big"
+)
+
+// A Bitset creates is an ordered list of bits with arbitrary length. It uses
```
Comment (lazynina) [inline utils/bitset/bitset.go:7]:
```suggestion
// A Bitset is an ordered list of bits with arbitrary length. It uses
```

File: utils/bitset/bitset.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+// the built in big.Int as the underlying storage scheme. The big.Int maintains
+// an ordered bit list and provides an interface where we can flip each bit
+// individually. The Bitset acts as a wrapper and provides boolean Get and Set
+// functions on top.
+//
+// We implement a custom Bitset data structure using Big.Int rather than using
+// an off-the-shelf solution because we need to support byte encoding and decoding,
+// with known endianness. Out of the box, the built in big.Int supports individual
```
Comment (lazynina) [inline utils/bitset/bitset.go:15]:
```suggestion
// with known endianness. Out of the box, the built-in big.Int supports individual
```

**Review by lazynina — APPROVED** [review]
functionality looks good - just some comment clean up

**Review by AeonSw4n — APPROVED** [review]
Nice and simple, I like it! And we're 💯 we won't fall victim to treacherous implementation details concealed in third-party libraries. I'm giving approval since I'm convinced the code works. Though I've left some comments around encode/decode. Could you address these before merging?

File: utils/bitset/bitset.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+func (b *Bitset) Size() int {
+	return b.store.BitLen()
+}
+
+// Return the absolute value of the underlying the BigSet as a big-endian
+// byte slice. The output is compressed such that if the underlying
+// big.Int had zeros at the highest bits, they will be removed
+// from the output.
+func (b *Bitset) ToBytes() []byte {
+	return b.store.Bytes()
```
[the remark this answers is not in the document]
Comment (tholonious) [reply utils/bitset/bitset.go:60]: Not in this case, and it's a subtle reason. When we call ToBytes() here, the length of the byte slice is captured in the slice data structure itself. It's up to the caller to call `len(bytes)` and do with the length as they please. By leaving the length out of the byte encoding, we make this struct minimally opinionated on how the length for the bytes is prefixed in network messages.

**Pull request merged**